Issue #332: Added option --addon-dir (defaults to the current work dir)

jsnajdr opened the pull request:
This solves #332: adds a new option to set the sources directory and override the process.cwd. This is an updated and polished commit from abandoned pull request #272

kumar303 commented on lib/utils.js:
```diff
@@ -78,16 +78,15 @@ exports.console = Object.freeze(jpmConsole);
  * @return {Object}
  */
-function getManifest () {
+function getManifest (addonDir) {
```
Can you make this more like `getManifest ({addonDir: ...})` instead? Because [this other patch](https://github.com/mozilla-jetpack/jpm/pull/394/files#diff-50e3aa130a4f97a42ee2cf111c7b1d9dR87) needs to pass in a new option to `getManifest()` as well.

kumar303 commented:
Can you also update the usage docs in the README ?

kumar303 commented:
Thanks for taking the time to re-do your patch, sounds like a good feature. I can help get it merged after the fix-ups.

kumar303 commented:
The patch from https://github.com/mozilla-jetpack/jpm/pull/394 just landed so you'll need to rebase on master for that. There should only be one conflict: I added `getManifest({xpiPath: ...})`. An easy way to do it would be to change your patch to use `getManifest ({addonDir: ...})` first, make sure the tests pass, then rebase and resolve the conflict.

jsnajdr commented:
Improved few things pointed out in review:
- arguments to getManifest are an options object with properties - compatible with the signing patch
- updated README.md documentation
- put back pushd/popd behavior into test before/after handlers
- fixed the sign command to work well with the addonDir option
- cleaned up the tests - return promise to Mocha instead of calling done()

One remaining issue:
- I could not push README.md that contained string "--addon-dir". The spell checking test detected it as spelling error. So I had to change it to "--dir". Should be fixed by someone more competent than me before merging.

kumar303 commented on test/functional/test.run.js:
```diff
@@ -288,7 +276,6 @@ describe("jpm run", function () {
       }
 
       it("run with only -v option", function(done) {
-        process.chdir(paramDumpPath);
```
after this, does the cmd need `{ addonDir: paramDumpPath ...` ? I can't quite tell what it's testing. I also noticed some `chdir(paramDumpPath)` calls are still in tests right after this but that's probably fine.

jsnajdr replied:
The other chdir's that are remaining in this file are useless and I'll remove them. Using the addonDir option is the right way. The tests are running the 'params-dump' addon in the fixtures directory and are testing its output.

kumar303 commented on test/unit/test.sign.js:
```diff
@@ -599,16 +599,13 @@ describe('amoClient.PseudoProgress', function() {
 
 describe('sign', function() {
   var simpleAddonPath = path.join(__dirname, "..", "fixtures", "simple-addon");
-  var manifest;
```
whoops, good catch

kumar303 commented on test/unit/test.utils.js:
```diff
@@ -29,18 +29,14 @@ describe("lib/utils", function () {
 
     it("returns manifest from custom XPI file", function () {
-      // Change into a non-package path.
-      process.chdir(path.join(__dirname, "..", "addons"));
```
This is a bit weird but what it did was made sure the test uses `simpleAddonXPI` rather than the add-on dir. So I think you should make this additional change :
````
return utils.getManifest({
    xpiPath: simpleAddonXPI, 
    addonDir: path.join(__dirname, "..", "addons"),  // pass in an invalid dir just to make sure it's not used
  })
    ...
````

kumar303 commented on test/utils/index.js:
```diff
@@ -54,7 +54,8 @@ function exec (args, options, callback) {
   var env = _.extend({}, options.env, process.env);
 
-  return child_process.exec("node " + jpm + " " + args, {
+  return child_process.exec("node " + jpm + " --addon-dir " + options.addonDir
```
what happens if `options.addonDir` is undefined? That could potentially break a test no? I think it would be a little safer as:
````
if (options.addonDir) {
  args = "--addon-dir " + options.addonDir + " " + args;
}
return child_process.exec("node " + jpm + " " + args ...
````

kumar303 commented:
Everything looks good! Thanks for the rapid follow-ups.

I'm on vacation next week but when I'm back (Nov 30) I'd like to check out your branch locally and just test it out a bit since it covers pretty much all jpm commands. Then I'll merge it in.

There were a couple signing bug fixes that just landed but nothing that will create conflicts. However, if you can rebase your changes on master then that would be helpful when I manually test it out.

kumar303 commented:
Thanks for your patience. I just tested it out locally and merged it into master. I only found one issue which was an easy fix: https://github.com/mozilla-jetpack/jpm/commit/af0859c5bc7585dbbec65a45866bb90d1cf5b824 Thanks again for the patch and spending so much time to get it in.

jsnajdr commented:
Thanks for the merge! Always happy to contribute to Firefox :)

danilvalov commented:
The package hasn't been updated in npm :(

kumar303 commented:
@danilvalov I just pushed a release https://www.npmjs.com/package/jpm Sorry about the delay.

danilvalov commented:
@kumar303 Thank you!
